lib/router-config/src/override_subgraph_urls.rs

```diff
@@ -0,0 +1,85 @@
+use std::collections::HashMap;
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+/// Configuration for how the Router should override subgraph URLs.
+/// This can be used to point to different subgraph endpoints based on environment,
+/// or to use dynamic expressions to determine the URL at runtime.
+#[derive(Debug, Default, Deserialize, Serialize, JsonSchema, Clone)]
+#[schemars(transparent, example = override_subgraph_urls_example_1())]
+#[serde(transparent)]
+pub struct OverrideSubgraphUrlsConfig(HashMap<String, PerSubgraphConfig>);
+
+impl OverrideSubgraphUrlsConfig {
+    pub fn get_subgraph_url(&self, subgraph_name: &str) -> Option<&UrlOrExpression> {
+        self.0.get(subgraph_name).map(|entry| &entry.url)
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone)]
+pub struct PerSubgraphConfig {
+    /// Overrides for the URL of the subgraph.
+    ///
+    /// For convenience, a plain string in your configuration will be treated as a static URL.
+    ///
+    /// ### Static URL Example
+    /// ```yaml
+    /// url: "https://api.example.com/graphql"
+    /// ```
+    ///
+    /// ### Dynamic Expression Example
+    ///
+    /// The expression has access to the following variables:
+    /// - `request`: The incoming HTTP request, including headers and other metadata.
+    /// - `original_url`: The original URL of the subgraph (from supergraph sdl).
+    ///
+    /// ```yaml
+    /// url:
+    ///   expression: |
+    ///     if .request.headers."x-region" == "us-east" {
+    ///       "https://products-us-east.example.com/graphql"
+    ///     } else if .request.headers."x-region" == "eu-west" {
+    ///       "https://products-eu-west.example.com/graphql"
+    ///     } else {
+    ///       .original_url
+    ///     }
+    pub url: UrlOrExpression,
+}
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone)]
+#[serde(untagged)]
+pub enum UrlOrExpression {
```

Comment:
I really like this way of configuring the different options between `String` and `{ ... }`. It makes it much simpler and easy to read and understand. Also feels more future-proof and we can add more stuff after (like subscriptions that we discussed).